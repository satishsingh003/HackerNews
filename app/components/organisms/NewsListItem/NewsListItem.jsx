import React from 'react';
import timeSince from '../../../utils/timeSince';

const NewsListItem = ({newsItem, hideNewsItem, upVoteNewsItem}) => {
    const hideThisItem = (e) => {
        e.preventDefault();
        hideNewsItem(newsItem);
    }

    const upVote = (e) => {
        e.preventDefault();
        if (!newsItem.upvoted) {
            upVoteNewsItem(newsItem);
        }
    }

    if (newsItem.title && !newsItem.hidden) {
        return <li key={newsItem.objectID}>
            <span className={"lgOnly"}>
                <span className={"commentsCount"}>{newsItem.num_comments}</span>
                <span className="upVote">
                    <span className={newsItem.upvoted ? 'upVoted' : ''}>{newsItem.points}</span>
                    <a href="#" onClick={upVote} className="upVoteIcon">Up</a>
                </span>
            </span>
            
            <span className="title space">{newsItem.title}</span>
            <br className="smOnly" />
            <span className="space">{`(${newsItem.url})`}</span>
            <br className="smOnly" />
            <span className="space author">{`by ${newsItem.author}`}</span>
            <span className="space timeSince">{timeSince(newsItem.created_at)}</span>
            <br className="smOnly" />
            <span className={"smOnly"}>
                <span className={"commentsCount"}>{newsItem.num_comments}</span>
                <span className="upVote">
                    <span className={newsItem.upvoted ? 'upVoted' : ''}>{newsItem.points}</span>
                    <a href="#" onClick={upVote} className="upVoteIcon">Up</a>
                </span>
            </span>
            <a href="#" className="space hideItem" onClick={hideThisItem}>[hide]</a>
            <style jsx>{`
                li {
                    padding: 3px 30px 3px 80px;
                    list-style: none;
                    margin: 5px 0;
                    font-size: 12px;
                }
                li:nth-child(even) {background: #CCC;}

                a {
                    text-decoration: none;
                    color: blue;
                    font-family: 'Arial';
                }

                a:hover {
                    opacity: 0.6;
                }

                .commentsCount, .upVote {
                    width: 70px;
                    display: inline-block;
                }
                .upVoteIcon {
                    width: 0;
                    height: 0;
                    border-style: solid;
                    border-width: 0 4px 8px 4px;
                    border-color: transparent transparent #000 transparent;
                    text-indent: -999px;
                    display: inline-block;
                    line-height: 10px;
                    margin-left: 10px;
                }
                .upVoted {
                    color: red;
                }
                .upVoted + .upVoteIcon {
                    border-color: transparent transparent #aaa !important;
                }
                .title {
                    font-size: 16px;
                }
                .timeSince {
                    color: #848484;
                }
                .author {
                    color: #4c4c4c;
                }
                .space {
                    margin-right: 10px;
                }
                .smOnly {
                    display: none;
                }
                @media only screen and (max-width: 1024px) {
                    li {
                        padding: 10px 10px;
                        line-height: 16px;
                    }
                    .smOnly {
                        display: block;
                    }
                    .lgOnly {
                        display: none;
                    }
                    .commentsCount {
                        width: 50px;
                    }
                    .upVote {
                        width: 60px;
                    }
                }
            `}</style>
        </li>
    }
    return null;
}

export default NewsListItem;