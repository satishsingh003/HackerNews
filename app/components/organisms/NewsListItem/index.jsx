import NewsListItem from './NewsListItem';
export default NewsListItem;