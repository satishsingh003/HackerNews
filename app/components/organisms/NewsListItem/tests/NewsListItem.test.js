import React from 'react';
import { shallow } from 'enzyme';
import NewsListItem from '../NewsListItem';

/*eslint-disable */
const propData = {
  newsItem: {
    created_at: '2017-11-28T19:41:10.000Z',
    title: 'macOS High Sierra: Anyone can login as “root” with empty password',
    url: 'https://twitter.com/lemiorhan/status/935578694541770752',
    author: 'vladikoff',
    points: 3001,
    story_text: null,
    comment_text: null,
    num_comments: 813,
    story_id: null,
    story_title: null,
    story_url: null,
    objectID: '15800676',
  },
  hideNewsItem: () => {},
  upVoteNewsItem: () => {},
};
/*eslint-enable */

describe('test the news list item component', () => {
  test('test snapshot', () => {
    const NewsListItemComponent = shallow(<NewsListItem {...propData} />);
    expect(NewsListItemComponent).toMatchSnapshot();
  });

  test('test if the rendered component has a li tag', () => {
    const NewsListItemComponent = shallow(<NewsListItem {...propData} />);
    expect(NewsListItemComponent.find('li')).toHaveLength(1);
  });

  test('test hide functionality if calling the props method', () => {
    const mockCallbackFunction = jest.fn();
    const newPropsData = { ...propData, hideNewsItem: mockCallbackFunction };
    const NewsListItemComponent = shallow(<NewsListItem {...newPropsData} />);
    NewsListItemComponent.find('.hideItem').simulate('click', {
      preventDefault: jest.fn,
    });
    expect(mockCallbackFunction.mock.calls.length).toBe(1);
  });
});
