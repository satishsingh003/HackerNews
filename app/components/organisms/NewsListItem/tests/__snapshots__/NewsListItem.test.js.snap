// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the news list item component test snapshot 1`] = `
<li
  key="15800676"
>
  <span
    className="lgOnly"
  >
    <span
      className="commentsCount"
    >
      813
    </span>
    <span
      className="upVote"
    >
      <span
        className=""
      >
        3001
      </span>
      <a
        className="upVoteIcon"
        href="#"
        onClick={[Function]}
      >
        Up
      </a>
    </span>
  </span>
  <span
    className="title space"
  >
    macOS High Sierra: Anyone can login as “root” with empty password
  </span>
  <br
    className="smOnly"
  />
  <span
    className="space"
  >
    (https://twitter.com/lemiorhan/status/935578694541770752)
  </span>
  <br
    className="smOnly"
  />
  <span
    className="space author"
  >
    by vladikoff
  </span>
  <span
    className="space timeSince"
  >
    2 years ago
  </span>
  <br
    className="smOnly"
  />
  <span
    className="smOnly"
  >
    <span
      className="commentsCount"
    >
      813
    </span>
    <span
      className="upVote"
    >
      <span
        className=""
      >
        3001
      </span>
      <a
        className="upVoteIcon"
        href="#"
        onClick={[Function]}
      >
        Up
      </a>
    </span>
  </span>
  <a
    className="space hideItem"
    href="#"
    onClick={[Function]}
  >
    [hide]
  </a>
  <style
    jsx={true}
  >
    
                li {
                    padding: 3px 30px 3px 80px;
                    list-style: none;
                    margin: 5px 0;
                    font-size: 12px;
                }
                li:nth-child(even) {background: #CCC;}

                a {
                    text-decoration: none;
                    color: blue;
                    font-family: 'Arial';
                }

                a:hover {
                    opacity: 0.6;
                }

                .commentsCount, .upVote {
                    width: 70px;
                    display: inline-block;
                }
                .upVoteIcon {
                    width: 0;
                    height: 0;
                    border-style: solid;
                    border-width: 0 4px 8px 4px;
                    border-color: transparent transparent #000 transparent;
                    text-indent: -999px;
                    display: inline-block;
                    line-height: 10px;
                    margin-left: 10px;
                }
                .upVoted {
                    color: red;
                }
                .upVoted + .upVoteIcon {
                    border-color: transparent transparent #aaa !important;
                }
                .title {
                    font-size: 16px;
                }
                .timeSince {
                    color: #848484;
                }
                .author {
                    color: #4c4c4c;
                }
                .space {
                    margin-right: 10px;
                }
                .smOnly {
                    display: none;
                }
                @media only screen and (max-width: 1024px) {
                    li {
                        padding: 10px 10px;
                        line-height: 16px;
                    }
                    .smOnly {
                        display: block;
                    }
                    .lgOnly {
                        display: none;
                    }
                    .commentsCount {
                        width: 50px;
                    }
                    .upVote {
                        width: 60px;
                    }
                }
            
  </style>
</li>
`;
